From 1955e4ab57eecd5619f4828e3714812b7d8d3b78 Mon Sep 17 00:00:00 2001
From: Marc Mao <marc.mao@intel.com>
Date: Thu, 6 Dec 2012 12:04:58 +0800
Subject: [PORT FROM MAIN] Hack FBO flush issue in captureScreen

BZ: 72011

In JB MR1, if hwc HAL version is bigger than 1.1. framebuffer HAL
will be closed by SurfaceFlinger. And SurfaceFlinger will not call
compositionComplete any more. This will cause captureScreen get some
corruption because IMG's OGL driver will not flush FBO context when
switch FBO in middle of a scene. This is a temp hack to this issue.
We add a compositionComplete interface to HWC hal and let it still
call it even HWC version is 1.1 later. If IMG provides a formal
solution for it, we will revert this patch.

Change-Id: I3c3d64fea889b34a239b492e3264b25fe5d949aa
Orig-Change-Id: I6da80ddcaff70af86dc48a9e27839b359a708967
Signed-off-by: Marc Mao <marc.mao@intel.com>
Reviewed-on: http://android.intel.com:8080/79390
Tested-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 .../surfaceflinger/DisplayHardware/HWComposer.cpp  |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/services/surfaceflinger/DisplayHardware/HWComposer.cpp b/services/surfaceflinger/DisplayHardware/HWComposer.cpp
index d8a0eb1..147fa38 100644
--- a/services/surfaceflinger/DisplayHardware/HWComposer.cpp
+++ b/services/surfaceflinger/DisplayHardware/HWComposer.cpp
@@ -752,8 +752,14 @@ int HWComposer::fbPost(int32_t id,
 }
 
 int HWComposer::fbCompositionComplete() {
-    if (mHwc && hwcHasApiVersion(mHwc, HWC_DEVICE_API_VERSION_1_1))
+    if (mHwc && hwcHasApiVersion(mHwc, HWC_DEVICE_API_VERSION_1_1)) {
+       if (mHwc->reserved_proc[0]) {
+            int (*compositionComplete)(struct hwc_composer_device_1*, int) =
+                (int (*)(hwc_composer_device_1*, int))mHwc->reserved_proc[0];
+            compositionComplete(mHwc, 0);
+        }
         return NO_ERROR;
+    }
 
     if (mFbDev->compositionComplete) {
         return mFbDev->compositionComplete(mFbDev);
-- 
1.7.4.1

